Update Clippy

flip1995 opened the pull request:
r? @Manishearth

smoelius committed:
`needless_borrow` uses `used_exactly_once`

WaffleLapkin committed:
Fix clippy tests that trigger `for_loop_over_fallibles` lint

camsteffen committed:
Introduce TypeErrCtxt

TypeErrCtxt optionally has a TypeckResults so that InferCtxt doesn't
need to.

compiler-errors committed:
Remove CastCheckResult since it's unused

camsteffen committed:
Change InferCtxtBuilder from enter to build

RalfJung committed:
make const_err a hard error

Urgau committed:
Stabilize half_open_range_patterns

smoelius committed:
Move some things around

smoelius committed:
Separate internal lints by pass

smoelius committed:
Expand `unnecessary_def_path` lint

smoelius committed:
Fix adjacent code

Noratrieb committed:
Add `as_ptr_cast_mut` lint

This lint detects calls to a `&self`-taking `as_ptr` method, where
the result is then immediately cast to a `*mut T`. Code like this
is probably invalid, as that pointer will not have write permissions,
and `*mut T` is usually used to write through.

Noratrieb committed:
Improve wording

Noratrieb committed:
Fix example

kraktus committed:
update rust version introduction

kraktus committed:
Add `manual_filter` lint for `Option`

Share much of its implementation with `manual_map` and should greatly benefit from its previous feedback.

kraktus committed:
refactor `manual_filter`

Move common functions to `manual_utils.rs`, better arm matching, use clippy utils `contains_unsafe_block`

roynrishingha committed:
add tests in `implicit_saturating_sub` lint

llogiq committed:
extend `box-default` lint, add suggestion

nyurik committed:
Fix to_string_in_format_args in parens

Fix suggestions like

```
print!("error: something failed at {}", (Location::caller().to_string()));
```

where the parenthesis enclose some portion of the value.

ouz-a committed:
Remove `mir::CastKind::Misc`

Alexendoo committed:
Replace if_chain with let chains in `clippy::author` output

flip1995 committed:
Merge commit 'ac0e10aa68325235069a842f47499852b2dee79e' into clippyup

flip1995 committed:
Merge commit '8f1ebdd18bdecc621f16baaf779898cc08cc2766' into clippyup

nyurik committed:
Change uninlined_format_args into a style lint

evantypanski committed:
Add curlies to scrutinees with side effects

evantypanski committed:
Fix flipped variable that made it through

bors committed:
Auto merge of #9601 - evantypanski:et/issue9575, r=Manishearth

[`match_single_binding`] Add curlies for more cases to fix suggestion causes error

Fixes #9575

changelog: [`match_single_binding`]: Add curlies for scrutinees with side effects for more cases

bors committed:
Auto merge of #9595 - Alexendoo:author-let-chains, r=Jarcho

Replace if_chain with let chains in `clippy::author` output

Should help nudge new contributors towards let chains

changelog: none

smoelius committed:
Fix adjacent code

bors committed:
Auto merge of #9585 - rust-lang:extend-box-default, r=Alexendoo

extend `box-default` lint, add suggestion

This extends the recently added `box-default` lint to also cover `Box::new(vec![])`, `Box::new(String::from(""))` and `Box::new(Vec::from([]))`. Also the lint now suggests a suitable replacement. I did not find a simple way to check whether the type is fully determined by the outside, so I at least checked for some variations to remove the turbofish in those cases.

---

changelog: none

WaffleLapkin committed:
fixup lint name

WaffleLapkin committed:
deprecate `clippy::for_loops_over_fallibles`

bors committed:
Auto merge of #102091 - RalfJung:const_err, r=oli-obk

make const_err a hard error

This lint has been deny-by-default with future incompat wording since [Rust 1.51](https://github.com/rust-lang/rust/pull/80394) and the stable release of this week starts showing it in cargo's future compat reports. I can't wait to finally get rid of at least some of the mess in our const-err-reporting-code. ;)

r? `@oli-obk`
Fixes https://github.com/rust-lang/rust/issues/71800
Fixes https://github.com/rust-lang/rust/issues/100114

nyurik committed:
Handle panic! inline_format-arg before ed2021

matthiaskrgr committed:
Rollup merge of #102675 - ouz-a:mir-technical-debt, r=oli-obk

Remove `mir::CastKind::Misc`

As discussed in #97649 `mir::CastKind::Misc` is not clear, this PR addresses that by creating a new enum variant for every valid cast.

r? ````@oli-obk````

kraktus committed:
[`unnecessary_cast`] Do not lint negative hexadecimal literals when cast as float

Floats cannot be expressed as hexadecimal literals

Jarcho committed:
Don't suggest moving tuple structs with a significant drop to late evaluation.

bors committed:
Auto merge of #9451 - kraktus:manual_filter2, r=dswij

Add `manual_filter` lint for `Option`

Share much of its implementation with `manual_map` and should greatly benefit from its previous feedback.
I'm sure it's possible to even more refactor both and would gladly take input on that as well as any clippy idiomatic usage, since this is my first lint addition.

I've added the lint to the complexity section for now, I don't know if every new lint needs to go in nursery first.

The matching could be expanded to more than `Some(<value>)` to lint on arbitrary struct matching inside the `Some` but I've left it like it was for `manual_map` for now. `needless_match::pat_same_as_expr` provides a more generic match example.

close https://github.com/rust-lang/rust-clippy/issues/8822

changelog: Add lint [`manual_filter`] for `Option`

bors committed:
Auto merge of #9386 - smoelius:further-enhance-needless-borrow, r=Jarcho

Further enhance `needless_borrow`, mildly refactor `redundant_clone`

This PR does the following:
* Moves some code from `redundant_clone` into a new `clippy_utils` module called `mir`, and wraps that code in a function called `dropped_without_further_use`.
* Relaxes the "is copyable" condition condition from #9136 by also suggesting to remove borrows from values dropped without further use. The changes involve the just mentioned function.
* Separates `redundant_clone` into modules.

Strictly speaking, the last bullet is independent of the others. `redundant_clone` is somewhat hairy, IMO. Separating it into modules makes it slightly less so, by helping to delineate what depends upon what.

I've tried to break everything up into digestible commits.

r? `@Jarcho`

(`@Jarcho` I hope you don't mind.)

changelog: continuation of #9136

llogiq committed:
add `cast-nan-to-int` lint

compiler-errors committed:
ImplItemKind::TyAlias => ImplItemKind::Type

smoelius committed:
Format affected files

unvalley committed:
feat: add Default to Lint groups

JohnTitor committed:
Rollup merge of #102829 - compiler-errors:rename-impl-item-kind, r=TaKO8Ki

rename `ImplItemKind::TyAlias` to `ImplItemKind::Type`

The naming of this variant seems inconsistent given that this is not really a "type alias", and the associated type variant for `TraitItemKind` is just called `Type`.

JohnTitor committed:
Rollup merge of #102275 - Urgau:stabilize-half_open_range_patterns, r=cjgillot

Stabilize `half_open_range_patterns`

This PR stabilize `feature(half_open_range_patterns)`:
```
Allows using `..=X` as a pattern.
```

And adds a new `feature(half_open_range_patterns_in_slices)` for the slice part, https://github.com/rust-lang/rust/pull/102275#issuecomment-1267422806.

The FCP was completed in https://github.com/rust-lang/rust/issues/67264.

compiler-errors committed:
Rename AssocItemKind::TyAlias to AssocItemKind::Type

Dylan-DPC committed:
Rollup merge of #99696 - WaffleLapkin:uplift, r=fee1-dead

Uplift `clippy::for_loops_over_fallibles` lint into rustc

This PR, as the title suggests, uplifts [`clippy::for_loops_over_fallibles`] lint into rustc. This lint warns for code like this:
```rust
for _ in Some(1) {}
for _ in Ok::<_, ()>(1) {}
```
i.e. directly iterating over `Option` and `Result` using `for` loop.

There are a number of suggestions that this PR adds (on top of what clippy suggested):
1. If the argument (? is there a better name for that expression) of a `for` loop is a `.next()` call, then we can suggest removing it (or rather replacing with `.by_ref()` to allow iterator being used later)
   ```rust
    for _ in iter.next() {}
    // turns into
    for _ in iter.by_ref() {}
    ```
2. (otherwise) We can suggest using `while let`, this is useful for non-iterator, iterator-like things like [async] channels
   ```rust
   for _ in rx.recv() {}
   // turns into
   while let Some(_) = rx.recv() {}
   ```
3. If the argument type is `Result<impl IntoIterator, _>` and the body has a `Result<_, _>` type, we can suggest using `?`
   ```rust
   for _ in f() {}
   // turns into
   for _ in f()? {}
   ```
4. To preserve the original behavior and clear intent, we can suggest using `if let`
   ```rust
   for _ in f() {}
   // turns into
   if let Some(_) = f() {}
   ```
(P.S. `Some` and `Ok` are interchangeable depending on the type)

I still feel that the lint wording/look is somewhat off, so I'll be happy to hear suggestions (on how to improve suggestions :D)!

Resolves #99272

[`clippy::for_loops_over_fallibles`]: https://rust-lang.github.io/rust-clippy/master/index.html#for_loops_over_fallibles

Dylan-DPC committed:
Rollup merge of #102868 - compiler-errors:rename-assoc-tyalias-to-ty, r=TaKO8Ki

Rename `AssocItemKind::TyAlias` to `AssocItemKind::Type`

Thanks `@camsteffen` for catching this in ast too, cc https://github.com/rust-lang/rust/pull/102829#issuecomment-1272649247

unvalley committed:
fix: change command order and remove waste condition

llogiq committed:
fix `box-default` ignoring trait objects' types

bors committed:
Auto merge of #9610 - Jarcho:fix-9608, r=Alexendoo

Don't suggest moving tuple structs with a significant drop to late evaluation

fixes #9608

changelog: Don't suggest moving tuple structs with a significant drop to late evaluation

GuillaumeGomez committed:
Fix unclosed HTML tag in clippy doc

Jarcho committed:
Use the correct type when comparing nested constants.

est31 committed:
Replace manual let else patterns with let else

bors committed:
Auto merge of #9572 - Nilstrieb:as-ptr-cast-mut, r=dswij

Add `as_ptr_cast_mut` lint

This lint detects calls to a `&self`-taking `as_ptr` method, where the result is then immediately cast to a `*mut T`. Code like this is probably invalid, as that pointer will not have write permissions, and `*mut T` is usually used to write through.

Examples of broken code with this pattern:
https://miri.saethlin.dev/ub?crate=lol_alloc&version=0.1.3
https://miri.saethlin.dev/ub?crate=sophon-wasm&version=0.19.0
https://miri.saethlin.dev/ub?crate=polars-core&version=0.24.2
https://miri.saethlin.dev/ub?crate=ach-cell&version=0.1.17

changelog: Add [`as_ptr_cast_mut`]

bors committed:
Auto merge of #9616 - unvalley:add-default-to-lint-groups, r=xFrednet

Add Default to Clippy Lints Lint groups

- related to #7958

This PR adds a default (reset) button to Clippy Lints Lint groups. (change for website)
[The page](https://rust-lang.github.io/rust-clippy/master/index.html) sets only `Deprecated` to false by default.
Certainly it is easy to set only `deprecated` to false, but it may be a bit lazy for beginners.

https://user-images.githubusercontent.com/38400669/194831117-3ade7e0d-c4de-4189-9daf-3be8ea3cdd18.mov

changelog: none

bors committed:
Auto merge of #9629 - est31:let_else, r=Jarcho

Replace manual let else patterns with let else

Clears the codebase from places where the lint added by #8437 is firing, by adopting let else.

changelog: none

smoelius committed:
Fix bug in `referent_used_exactly_once`

bors committed:
Auto merge of #9605 - nyurik:fix-inline-edition, r=llogiq

fix: uninlined_format_args shouldn't inline panic! before 2021ed

Before 2021 edition, `panic!("...")` was not treated as a format string.
Clippy autofix of `panic!("{}", foo)` into `panic!("{foo}")` is incorrect.

changelog: [`uninlined_format_args`]: Do not inline panic! macros before 2021 edition

kraktus committed:
`default_numeric_fallback` do not lint on constants

roynrishingha committed:
update Applicability of `rc_buffer` from `MachineApplicable` to
`Unspecified`

bors committed:
Auto merge of #9633 - royrustdev:fix_rc_buffer, r=llogiq

update Applicability of `rc_buffer` lint from `MachineApplicable` to `Unspecified`

`Unspecified`

This changes `rc_buffer` from MachineApplicable to Unspecified

```
changelog: change [`rc_buffer`] to Unspecified.
```

fixes #6241

---

changelog: change [`rc_buffer`] to Unspecified.

bors committed:
Auto merge of #9617 - llogiq:cast-nan-to-int, r=Alexendoo

add `cast-nan-to-int` lint

This fixes #371.

r? `@Alexendoo`

---

changelog: add [`cast-nan-to-int`] lint

roynrishingha committed:
Update Applicability of `redundant_allocation` lint from
`MachineApplicable` to `Unspecified`

bors committed:
Auto merge of #9627 - Jarcho:ice-9625, r=xFrednet

Use the correct type when comparing nested constants.

fixes #9625

changelog: `manual_range_contains`: fix ICE when the values are behind a reference

kraktus committed:
refactor `default_numeric_fallback`

We only need to store if the literal binding has an explicit type bound or not

bors committed:
Auto merge of #9590 - nyurik:fix-parens, r=Alexendoo

Fix to_string_in_format_args in parens

Fix suggestions like

```
print!("error: something failed at {}", (Location::caller().to_string()));
```

where the parenthesis enclose some portion of the value.

Fixes #9540

changelog: [`to_string_in_format_args`]: fix incorrect fix when value is enclosed in parenthesis

Alexendoo committed:
Fix edition revision ui tests

bors committed:
Auto merge of #9600 - nyurik:inline-fmt-style, r=llogiq

Change uninlined_format_args into a style lint

As [previously discussed](https://github.com/rust-lang/rust-clippy/pull/9233#issuecomment-1256361205), the `uninlined_format_args` should probably be a part of the default style because `println!("{}", foo)` is not as concise or easy to understand as `println!("{foo}")`

changelog: [`uninlined_format_args`]: change to be the default `style`

bors committed:
Auto merge of #9584 - royrustdev:implicit_saturating_sub, r=llogiq

add tests in `implicit_saturating_sub` lint

This adds more tests to the `implicit_saturating_sub` lint to rule out certain false positives that have appeared in the past.

Now with those false positives out of the equation, we can move the lint to `style`.

---

changelog: promote [`implicit-saturating-sub`] to the `style` category

Alexendoo committed:
Add `unused_format_specs` lint

bors committed:
Auto merge of #9640 - Alexendoo:edition-revisions, r=llogiq

Fix edition revision ui tests

#9605 had me wondering how the edition revision tests were working for `manual_assert` but not for `@nyurik,` but it turns out `manual_assert`'s tests weren't working either. I checked how `rust-lang/rust` does it and apparently it comes down to whitespace, `//[rev] edition:X` works 😬

Removes the revisions from `match_wild_err_arm` as I couldn't find any edition dependant behaviour there

r? `@llogiq`

changelog: none

Alexendoo committed:
Enable cargo sparse registry in CI

bors committed:
Auto merge of #9641 - Alexendoo:sparse-registry, r=flip1995

Enable cargo sparse registry in CI

https://blog.rust-lang.org/2022/06/22/sparse-registry-testing.html

The initial registry update takes around 1 minute currently, so this gives quite a nice speed boost to CI build times

r? `@flip1995`

changelog: none

Commit:
Book: Small grammar + link a11y change

hkBst committed:
add missing comma

bors committed:
Auto merge of #9644 - hkBst:patch-1, r=flip1995

add missing comma

changelog: none

bors committed:
Auto merge of #9643 - icecream17:patch-1, r=flip1995

Book: Small grammar + link a11y change

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: none

---

Very minor

For the link accessibility change, `here` and related don't provide context for screen readers who are reading a list of links.
(Random supporting google links)
https://www.w3.org/QA/Tips/noClickHere
https://usability.yale.edu/web-accessibility/articles/links

Jarcho committed:
Don't lint `ptr_arg` when used as an incompatible trait object

bors committed:
Auto merge of #9645 - Jarcho:ptr_arg_9542, r=llogiq

Don't lint `ptr_arg` when used as an incompatible trait object

fixes #9542
changelog: [`ptr_arg`](https://rust-lang.github.io/rust-clippy/master/#ptr_arg): Don't lint when used as an incompatible trait object

Alexendoo committed:
Add a suggestion and a note about orphan rules for `from_over_into`

bors committed:
Auto merge of #9649 - Alexendoo:from-over-into-suggestion, r=llogiq

Add a suggestion and a note about orphan rules for `from_over_into`

Adds a machine applicable suggestion to convert the `Into` impl into a `From` one to `from_over_into`

Also adds a note explaining that `impl From<Local> for Foreign` is fine if the `Into` type is foreign

Closes #7444
Addresses half of #9638

changelog: [`from_over_into`] Add a suggestion and a note about orphan rules

kraktus committed:
`explicit_ty_bound` code golf

kraktus committed:
[`zero_prefixed_literal`] Do not advise to use octal form if not possible

llogiq committed:
fix `box-default` linting `no_std` non-boxes

bors committed:
Auto merge of #9655 - llogiq:unbox-default, r=dswij

fix `box-default` linting `no_std` non-boxes

This fixes #9653 by doing the check against the `Box` type correctly even if `Box` isn't there, as in `no_std` code. Thanks to `@lukas-code` for opening the issue and supplying a reproducer!

---

changelog: none

bors committed:
Auto merge of #9636 - kraktus:numeric-fallback, r=dswij

[`default_numeric_fallback`] do not lint on constants

fix https://github.com/rust-lang/rust-clippy/issues/9632

changelog:[`default_numeric_fallback`] do not lint on constants

bors committed:
Auto merge of #9566 - smoelius:diagnostic-item-path, r=dswij

Expand internal lint `unnecessary_def_path`

This PR does essentially two things:
* Separates the internal lints into modules by pass. (`internal_lints.rs` was over 1400 lines, which is a little unruly IMHO.)
* ~Adds a new~ Expands the `unnecessary_def_path` internal lint to flag hardcoded paths to diagnostic and language items.

My understanding is that the latter is currently done by reviewers. Automating this process should make things easier for both reviewers and contributors.

I could make the first bullet a separate PR, or remove it entirely, if desired.

changelog: Add internal lint `diagnostic_item_path`

TennyZhuang committed:
Add new lint `partial_pub_fields`

Signed-off-by: TennyZhuang <zty0826@gmail.com>

TennyZhuang committed:
fix dogfood

Signed-off-by: TennyZhuang <zty0826@gmail.com>

TennyZhuang committed:
add many tests

Signed-off-by: TennyZhuang <zty0826@gmail.com>

TennyZhuang committed:
fix a doctest

Signed-off-by: TennyZhuang <zty0826@gmail.com>

bors committed:
Auto merge of #9658 - TennyZhuang:partial-pub-fields, r=llogiq

Add new lint `partial_pub_fields`

Signed-off-by: TennyZhuang <zty0826@gmail.com>

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: `partial_pub_fields`: new lint to disallow partial fields of a struct be pub

Resolve #9604

bors committed:
Auto merge of #9652 - kraktus:octo_89, r=xFrednet

[`zero_prefixed_literal`] Do not advise to use octal form if not possible

fix https://github.com/rust-lang/rust-clippy/issues/9651

changelog: [`zero_prefixed_literal`] Do not advise to use octal form if not possible

bors committed:
Auto merge of #9609 - kraktus:hexa_f32, r=giraffate

[`unnecessary_cast`] Do not lint negative hexadecimal literals when cast as floats

fix https://github.com/rust-lang/rust-clippy/issues/9603

changelog: [`unnecessary_cast`] Do not lint negative hexadecimal literals when cast as floats

bors committed:
Auto merge of #9637 - Alexendoo:unused-format-specs, r=xFrednet

Add `unused_format_specs` lint

Currently catches two cases:

An empty precision specifier:

```rust
// the same as {}
println!("{:.}", x);
```

And using formatting specs on `format_args!()`:

```rust
// prints `x.`, not `x    .`
println("{:5}.", format_args!("x"));
```

changelog: new lint: [`unused_format_specs`]

Alexendoo committed:
Add `missing_trait_methods` lint

bors committed:
Auto merge of #9634 - royrustdev:fix_redundant_allocation, r=Manishearth

Update Applicability of `redundant_allocation` lint from `MachineApplicable` to `MaybeIncorrect`

This changes `redundant_allocation` lint from MachineApplicable to MaybeIncorrect

```
changelog: [`redundant_allocation]: Change Applicability from MachineApplicable to MaybeIncorrect

```

fixes #6243

---
changelog: [`redundant_allocation`]: Change Applicability from MachineApplicable to MaybeIncorrect

flip1995 committed:
Merge remote-tracking branch 'upstream/master' into rustup

flip1995 committed:
Bump nightly version -> 2022-10-20

bors committed:
Auto merge of #9680 - flip1995:rustup, r=flip1995

Rustup

r? `@ghost`

I ran out of time and will have to do the Clippy->Rust sync tomorrow.

changelog: none

bors committed:
Auto merge of #9670 - Alexendoo:missing-trait-methods, r=Jarcho

Add `missing_trait_methods` lint

Closes #9661

changelog: new lint: [`missing_trait_methods`]

bors committed:
Auto merge of #9635 - smoelius:fix-9386-bug, r=Jarcho

Fix bug introduced by #9386

#9386 introduced a potential out-of-bounds array access. Specifically, a location returned by `local_assignments` could have  [`location.statement_index` equal to `mir.basic_blocks[location.block].statements.len()`](https://github.com/rust-lang/rust-clippy/blob/b8a9a507bf9e3149d287841454842116c72d66c4/clippy_utils/src/mir/mod.rs#L129), in which case the location would refer to the block terminator:
https://github.com/rust-lang/rust-clippy/blob/b8a9a507bf9e3149d287841454842116c72d66c4/clippy_lints/src/dereference.rs#L1204-L1206
I suspect the bug is not triggerable now, because of checks leading up to where it occurs. But a future code change could make it triggerable. Hence, it should be fixed.

r? `@Jarcho`

changelog: none

kraktus committed:
`ref_option_ref` do not lint when inner reference is mutable

As it makes the `Option` Non Copy

kraktus committed:
[`collapsible_match`] specify field name when destructuring structs

bors commented:
:umbrella: The latest upstream changes (presumably #103310) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Manishearth commented:
argh merge conflict

bors committed:
Auto merge of #9685 - kraktus:collapsible_match, r=llogiq

[`collapsible_match`] specify field name when destructuring structs

changelog: [`collapsible_match`] specify field name when destructuring structs

fix https://github.com/rust-lang/rust-clippy/issues/9647

I wasn't the sure about the best way to convey the message in the lint message since it does not use suggestion. Because I liked the former output highlighting both spans, I've left it as before, only modifying the span label.

bors committed:
Auto merge of #9684 - kraktus:ref_option_ref, r=xFrednet

`ref_option_ref` do not lint when inner reference is mutable

changelog: FP: [`ref_option_ref`]: No longer lints if the inner reference is mutable

fix https://github.com/rust-lang/rust-clippy/issues/9682

Alexendoo committed:
Move MSRV tests into the lint specific test files

smoelius committed:
Apply manual fixes

smoelius committed:
Add `lintcheck` to packages linted by `dogfood` test

smoelius committed:
Apply `--fix` fixes

bors committed:
Auto merge of #9691 - smoelius:lint-lintcheck, r=llogiq

Add `lintcheck` to packages linted by `dogfood` test

Currently, `lintcheck` is not checked by the `dogfood` test. I assume that is not intentional. If it is intentional, then this PR can be ignored.

changelog: Add `lintcheck` to packages linted by `dogfood` test

bors committed:
Auto merge of #9688 - Alexendoo:msrv-tests, r=Manishearth

Move MSRV tests into the lint specific test files

There are currently two ways MSRV tests are done in the ui test suite, adding a case to the `#![clippy::msrv = "1.0"]` `tests/ui/min_rust_version_attr.rs` or adding the two `msrv_1_xx` functions to the test file of the lint in question

This updates the clippy book to suggest the `msrv_1_xx` style, and replaces the tests in `tests/ui/min_rust_version_attr.rs` with ones of that style

Almost the entire diff is just moving stuff around as a result, I made sure to check the line numbers the lints are emitted at correspond with the right `msrv` case, so feel free to only skim that part

changelog: none

bors committed:
Auto merge of #9622 - llogiq:box-dyn-default, r=Alexendoo

fix `box-default` ignoring trait objects' types

This avoids removing the turbofish when the `Box` type is a `dyn` or `impl _`.

This fixes #9621.

---

changelog: none

flip1995 committed:
Merge commit '4f142aa1058f14f153f8bfd2d82f04ddb9982388' into clippyup

flip1995 commented:
re-synced @Manishearth

Manishearth commented:
@bors r+ p=1

bors commented:
:pushpin: Commit da9755b6ba108b04eb9dfc771d0a6dd2f86f911b has been approved by `Manishearth`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Manishearth da9755b6ba108b04eb9dfc771d0a6dd2f86f911b -->
<!-- homu: {"type":"Approved","sha":"da9755b6ba108b04eb9dfc771d0a6dd2f86f911b","approver":"Manishearth","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit da9755b6ba108b04eb9dfc771d0a6dd2f86f911b with merge 758f19645b8ebce61ea52d1f6672fd057bc8dbee...
<!-- homu: {"type":"BuildStarted","head_sha":"da9755b6ba108b04eb9dfc771d0a6dd2f86f911b","merge_sha":"758f19645b8ebce61ea52d1f6672fd057bc8dbee"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3314589431/jobs/5476346419)
Approved by: Manishearth
Pushing 758f19645b8ebce61ea52d1f6672fd057bc8dbee to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Manishearth","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3314589431/jobs/5476346419"},"merge_sha":"758f19645b8ebce61ea52d1f6672fd057bc8dbee"} -->

rust-timer commented:
Finished benchmarking commit (758f19645b8ebce61ea52d1f6672fd057bc8dbee): [comparison URL](https://perf.rust-lang.org/compare.html?start=c07a6a9c0c6c571d535de74d7ec28469fba46e37&end=758f19645b8ebce61ea52d1f6672fd057bc8dbee&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=c07a6a9c0c6c571d535de74d7ec28469fba46e37&end=758f19645b8ebce61ea52d1f6672fd057bc8dbee&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c07a6a9c0c6c571d535de74d7ec28469fba46e37&end=758f19645b8ebce61ea52d1f6672fd057bc8dbee&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 4.2%     | [2.6%, 5.6%] | 3         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -3.0%    | [-3.6%, -2.5%] | 2         |
| All ❌✅ (primary) | -        | -     | 0         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=c07a6a9c0c6c571d535de74d7ec28469fba46e37&end=758f19645b8ebce61ea52d1f6672fd057bc8dbee&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*
